File: src/root-query/parse-root-query-ast.ts
import { parseDirectionToken } from "./parse-direction-token.js";
import { parseEndpointQuery } from "./parse-endpoint-query.js";
import type { ParseResult } from "./parse-error.js";
import type { RootQueryScope } from "./root-query-scope.js";
import type { RootQuery } from "./root-query.js";

export function parseRootQueryAst(
  token: string,
  scope: RootQueryScope,
): ParseResult<RootQuery> {
  if (token === "") {
    return { ok: false, errors: [{ message: "empty root query" }] };
  }

  const parts = token.split("..");

  if (parts.length >= 3) {
    return {
      ok: false,
      errors: [{ message: `unexpected duplicate '..' in '${token}'` }],
    };
  }

  if (parts.length === 1) {
    const lone = parts[0] ?? "";
    const r = parseEndpointQuery(lone);
    if (!r.ok) {
      return r;
    }
    if (!scope.point) {
      return {
        ok: false,
        errors: [{ message: `unexpected token '${token}'` }],
      };
    }
    return {
      ok: true,
      value: { kind: "single", query: r.value, raw: token },
    };
  }

  const lhsText = parts[0] ?? "";
  const rhsText = parts[1] ?? "";

  if (lhsText === "") {
    return {
      ok: false,
      errors: [
        { message: `unexpected empty left-hand side of '..' in '${token}'` },
      ],
    };
  }
  if (rhsText === "") {
    return {
      ok: false,
      errors: [
        { message: `unexpected empty right-hand side of '..' in '${token}'` },
      ],
    };
  }

  const lhsR = parseEndpointQuery(lhsText);
  if (!lhsR.ok) {
    return lhsR;
  }

  if (rhsText.startsWith("+")) {
    const dirR = parseDirectionToken(rhsText);
    if (!dirR.ok) {
      return dirR;
    }
    if (!scope.direction) {
      return {
        ok: false,
        errors: [
          { message: `unexpected direction token '${rhsText}' in '${token}'` },
        ],
      };
    }
    if (dirR.value.level !== null && !scope.directionLevel) {
      return {
        ok: false,
        errors: [
          {
            message: `unexpected level in direction token '${rhsText}'`,
          },
        ],
      };
    }
    return {
      ok: true,
      value: {
        kind: "direction",
        lhs: lhsR.value,
        dir: dirR.value.dir,
        level: dirR.value.level,
        raw: token,
      },
    };
  }

  const rhsR = parseEndpointQuery(rhsText);
  if (!rhsR.ok) {
    return rhsR;
  }
  if (!scope.path) {
    return {
      ok: false,
      errors: [{ message: `unexpected '..' in '${token}'` }],
    };
  }
  return {
    ok: true,
    value: {
      kind: "path",
      lhs: lhsR.value,
      rhs: rhsR.value,
      raw: token,
    },
  };
}

File: crates/unsnarl-root-query/src/parse_root_query_ast.rs
use crate::parse_direction_token::parse_direction_token;
use crate::parse_endpoint_query::parse_endpoint_query;
use crate::parse_error::ParseError;
use crate::root_query::RootQuery;
use crate::root_query_scope::RootQueryScope;

pub fn parse_root_query_ast(
    token: &str,
    scope: &RootQueryScope,
) -> Result<RootQuery, Vec<ParseError>> {
    if token.is_empty() {
        return Err(vec![ParseError {
            message: "empty root query".to_string(),
        }]);
    }

    let mut parts = token.splitn(3, "..");
    let first = parts
        .next()
        .expect("splitn always yields at least one item");
    let second = parts.next();
    let third = parts.next();

    if third.is_some() {
        return Err(vec![ParseError {
            message: format!("unexpected duplicate '..' in '{token}'"),
        }]);
    }

    let Some(rhs_text) = second else {
        let query = parse_endpoint_query(first)?;
        if !scope.point {
            return Err(vec![ParseError {
                message: format!("unexpected token '{token}'"),
            }]);
        }
        return Ok(RootQuery::Single {
            query,
            raw: token.to_string(),
        });
    };

    let lhs_text = first;

    if lhs_text.is_empty() {
        return Err(vec![ParseError {
            message: format!("unexpected empty left-hand side of '..' in '{token}'"),
        }]);
    }
    if rhs_text.is_empty() {
        return Err(vec![ParseError {
            message: format!("unexpected empty right-hand side of '..' in '{token}'"),
        }]);
    }

    let lhs = parse_endpoint_query(lhs_text)?;

    if rhs_text.starts_with('+') {
        let dir_tok = parse_direction_token(rhs_text)?;
        if !scope.direction {
            return Err(vec![ParseError {
                message: format!("unexpected direction token '{rhs_text}' in '{token}'"),
            }]);
        }
        if dir_tok.level.is_some() && !scope.direction_level {
            return Err(vec![ParseError {
                message: format!("unexpected level in direction token '{rhs_text}'"),
            }]);
        }
        return Ok(RootQuery::Direction {
            lhs,
            dir: dir_tok.dir,
            level: dir_tok.level,
            raw: token.to_string(),
        });
    }

    let rhs = parse_endpoint_query(rhs_text)?;
    if !scope.path {
        return Err(vec![ParseError {
            message: format!("unexpected '..' in '{token}'"),
        }]);
    }
    Ok(RootQuery::Path {
        lhs,
        rhs,
        raw: token.to_string(),
    })
}

#[cfg(test)]
#[path = "parse_root_query_ast_test.rs"]
mod parse_root_query_ast_test;
